### ISSUE TITLE
ML-based commercial detection (optional classifier)

### ISSUE THREAD

**Issue by AndyP2** [erikkaashoek/Comskip#200]
I have this idea but don’t currently have the time to implement it myself.  
These notes (developed with help from Perplexity) describe one possible way to add an optional ML classifier to comskip.

I’m opening this as a discussion point and design sketch in case:
- The maintainers think it’s worth doing eventually, and
- Someone else (now or in future) would like a concrete project to work on.

I’m happy for anyone to take this on; if I get time later, I may also contribute, but I've got a lot on at the moment so may not get to it any time soon.

# ML-based commercial detection (optional classifier)

## Summary

Add an optional machine-learning classifier as a new detection method in comskip, so that ad vs content decisions can be learned from data instead of relying solely on hand-tuned thresholds and heuristics.

The goal is to:
- Improve detection quality (fewer false positives on content, better handling of edge cases).
- Keep full backwards compatibility: existing behaviour unchanged unless the user explicitly enables ML.
- Allow users to train their own models on their own channels/recordings.

## Motivation

comskip already computes rich per-frame and per-segment features (brightness, volume, logo presence, scene changes, aspect ratio, etc.) and combines them using weighted rules and thresholds (see example `comskip.ini` files in the [PlexComskip](https://github.com/ekim1337/PlexComskip/blob/master/comskip.ini) and [etv-comskip](https://github.com/essandess/etv-comskip/blob/master/src/comskip_ini/comskip.ini) repos, and the [tuning guide](http://www.kaashoek.com/gbpvr/tuning.htm) on kaashoek.com).

A learned classifier can:
- Combine these features in a non-linear way.
- Adapt to specific channels, regions, or recording setups.
- Reduce the need for manual tuning of many interdependent thresholds.

This is particularly useful when:
- Channels have subtle or inconsistent ad markers.
- Standard heuristics occasionally mark content as ads (the most costly error for users).

### New detection method

Introduce a new bit flag, e.g.:

```c
#define DETECT_ML 256
```

Used in `comskip.ini` like existing methods:

```ini
detect_method=255   ; existing: all current methods
detect_method=511   ; existing + ML (255 + 256)
```

### New [ML] section in comskip.ini

Example configuration:

```ini
[ML]
; Path to the ONNX model (relative or absolute)
ml_model_path=C:\models\comskip_ml.onnx

; Probability threshold above which a segment is considered an ad
ml_threshold=0.5

; If 1, combine ML score with existing heuristic score; if 0, use ML only
ml_use_heuristics=1

; Relative weight of ML score vs heuristic score when ml_use_heuristics=1
ml_weight=1.0
```

Defaults:
- ML disabled unless `detect_method` includes `256`.
- If enabled but model path is invalid, comskip should log an error and fall back to non-ML behaviour (or fail gracefully, depending on project policy).

### Feature set

Use features comskip already computes, aggregated per segment (between cutpoints). A minimal initial set:

- `mean_brightness`
- `max_brightness`
- `brightness_variance`
- `mean_volume`
- `max_volume`
- `silence_ratio`
- `logo_presence_ratio`
- `avg_logo_confidence`
- `scene_changes_per_sec`
- `aspect_ratio_changes`
- `segment_duration_sec`
- Optional: `distance_to_nearest_black_frame`, `uniform_frame_ratio`

These map directly to existing comskip metrics and CSV outputs (see the [tuning guide](http://www.kaashoek.com/gbpvr/tuning.htm) and example ini files linked above).

### Model format

Recommend **ONNX** as the model format:

- Widely supported, stable C/C++ API (ONNX Runtime).
- Easy to train in many frameworks (PyTorch, TensorFlow, scikit-learn, ML.NET, etc.) and export.
- No need to embed a heavy framework in comskip; just link ONNX Runtime.

The comskip binary would:
- Load the ONNX model at startup (if ML is enabled).
- For each segment, assemble the feature vector and call `ml_predict(features) -> probability`.
- Convert probability to a score and combine with existing heuristic scores.

### Scoring integration

When `detect_method & DETECT_ML`:

1. Compute ML probability `p` for each segment.
2. Derive an ML score, e.g.:
   - `ml_score = p` (continuous), or
   - `ml_score = (p >= ml_threshold) ? 1.0f : 0.0f`.
3. Combine with existing heuristic score:
   - If `ml_use_heuristics == 0`: `final_score = ml_score`.
   - If `ml_use_heuristics == 1`:

     ```text
     final_score = (heuristic_score + ml_weight * ml_score) / (1 + ml_weight)
     ```

4. Pass `final_score` into the existing thresholding and post-processing logic (min/max ad length, merging, etc.).

This keeps the rest of comskip’s pipeline unchanged.

## Training workflow (for users)

Outline a simple workflow for users who want to train their own models:

1. Run comskip on a set of recordings with CSV output enabled.
2. Manually correct the generated cutlists to obtain ground-truth ad/content segments.
3. Use a small external tool (Python/C#/etc.) to:
   - Parse the CSV.
   - Aggregate per-frame metrics into per-segment features.
   - Export a training CSV with features + binary labels.
4. Train a small model (e.g. 1–2 layer MLP) in any framework and export to ONNX.
5. Configure `comskip.ini` to point to the new model and enable `DETECT_ML`.

The project could optionally provide:
- A sample training script (Python or similar).
- A small example model trained on a generic dataset.

## Implementation notes

- Link ONNX Runtime as an optional dependency; ML code only compiled/used if enabled.
- Keep ML code isolated:
  - `ml_init(const char* model_path)`
  - `ml_predict(const float* features, int num_features, float* out_prob)`
  - `ml_cleanup()`
- Ensure no behaviour change when ML is disabled.
- Provide clear error messages if:
  - Model file is missing or invalid.
  - Feature vector size does not match model input.

## Benefits

- Better ad detection quality, especially on tricky channels.
- Reduced need for users to manually tune many thresholds.
- Extensible: more features or more sophisticated models can be added later without changing the core logic.

## Open questions

- Should the project provide a reference model or just the infrastructure?
- What is the minimum acceptable model size/performance for a “reference” model?
- Should ML be built by default, or behind a CMake/Makefile flag (e.g. `-DENABLE_ML=ON`)?

## Next steps

- Discuss design and priorities with maintainers.
- If agreed, implement a minimal version:
  - Ini parsing + config.
  - ONNX Runtime integration.
  - Basic ML scoring combined with existing heuristics.
- Later iterations can improve:
  - Feature set.
  - Model architecture.
  - Training tooling and documentation.
